drafts/str4d-overwinter-sighash/draft1.rst

```diff
@@ -0,0 +1,312 @@
+The transaction digest is additionally used for the JoinSplit signature (solely with sighash type ``ALL``).
+[#zcash-protocol]_
+
+Unfortunately, there are at least 2 weaknesses in the original SignatureHash transaction digest algorithm:
+
+* For the verification of each signature, the amount of data hashing is proportional to the size of the
+  transaction. Therefore, data hashing grows in O(n\ :sup:`2`) as the number of sigops in a transaction
+  increases. While a 1 MB block would normally take 2 seconds to verify with an average computer in 2015, a
+  1MB transaction with 5569 sigops may take 25 seconds to verify. This could be fixed by optimizing the digest
+  algorithm by introducing some reusable “midstate”, so the time complexity becomes O(n). [#quadratic]_
+
+* The algorithm does not involve the value being spent by the input. This is usually not a problem for online
+  network nodes as they could request for the specified transaction to acquire the output value. For an
+  offline transaction signing device ("cold wallet"), however, the lack of knowledge of input amount makes it
+  impossible to calculate the exact amount being spent and the transaction fee. To cope with this problem a
+  cold wallet must also acquire the full transaction being spent, which could be a big obstacle in the
+  implementation of a lightweight, air-gapped wallet. By including the input value of part of the transaction
+  digest, a cold wallet may safely sign a transaction by learning the value from an untrusted source. In the
+  case that a wrong value is provided and signed, the signature would be invalid and no funds would be lost.
+  [#offline-wallets]_
+
+Deploying the aforementioned fixes in the original script system is not a simple task.
+
+
+Specification
+=============
+
+A new transaction digest algorithm is defined, but only applicable from the Overwinter upgrade block height
+[#ZIP0000]_::
+  BLAKE2b-256 of the serialization of:
+    1. nVersion of the transaction (4-byte little endian)
+    2. nVersionBranch of the transaction (4-byte little endian)
+    3. hashPrevouts (32-byte hash)
+    4. hashSequence (32-byte hash)
+    5. hashOutputs (32-byte hash)
+    6. hashJoinSplits (32-byte hash)
+    7. nLockTime of the transaction (8-byte little endian)
+    8. nExpiryTime of the transaction (8-byte little endian)
+    9. sighash type of the signature (4-byte little endian)
+   10. If we are serializing an input (ie. this is not a JoinSplit signature hash):
+       a. outpoint (32-byte hash + 4-byte little endian) 
+       b. scriptCode of the input (serialized as scripts inside CTxOuts) [TODO]
+       c. value of the output spent by this input (8-byte little endian)
+       d. nSequence of the input (4-byte little endian)
+
+The BLAKE2b-256 personalization field is set to::
+
+  "ZcashSigHash" || BLOCK_BRANCH_ID
+
+``BLOCK_BRANCH_ID`` is the ``BRANCH_ID`` for the epoch of the transaction's parent block. [#ZIP-activation-mechanism]_
```

Comment:
Can we think of terminology that better clarifies the different identifiers? e.g. FAMILY_BRANCH_ID (I think family, e.g. Zcash or some clone) whereas EPOCH_ID (epoch makes me think of time so then I start thinking that the ID will expire, makes me less confident/comfortable for some reason).

Reply:
Above comment is probably not relevant given recent discussion.

Reply:
Decided to name it `CONSENSUS_BRANCH_ID` to match usage in the code.